resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala

```diff
@@ -462,6 +462,31 @@ package object config extends Logging {
     .stringConf
     .createWithDefault("yarn.io/fpga")
 
+  private[spark] val YARN_CLIENT_STAT_CACHE_PRELOADED_ENABLED =
+    ConfigBuilder("spark.yarn.client.statCache.preloaded.enabled")
+    .doc("This configuration enables statCache to be preloaded at YARN client side. This feature " +
+      "analyzes the pattern of resources paths. If multiple resources shared the same parent " +
+      "directory, a single <code>listStatus</code> will be invoked on the parent directory " +
+      "instead of multiple <code>getFileStatus</code> performed on each individual resources. " +
+      "If most resources are from the same directory, this feature greatly reduces the RPC call, " +
+      "reduce job runtime, and improve the job runtime variation due to network delays. " +
```

Comment:
We won't actually decrease the runtime of the job itself, just of the submission/initialization of the job. Generally I think we should be careful about making too strong of a claim. How about "If most resources from a small set of directories, this can substantially improve job submission time."